### ISSUE TITLE
Potential eengine bug: Functions and procedures without parameters

### ISSUE THREAD

**Issue by ronaldtse** [lutaml/expressir#23]
Functions and procedures without parameters

Declaration:

```
PROCEDURE emptyProcedure; END_PROCEDURE;
```

Call:

```
emptyProcedure;
```

eengine throws `Error: ASSIGNMENT_STMT expected ':COLON-EQUAL', got '#\;'`

It seems that the spec doesn't explicitly disallow them, or at least I couldn't find where.

"A function or procedure **can** have formal parameters." in section 9.5.3

BNF syntax has parameters in `[` `]`, which means optional

```
function_head = FUNCTION function_id [ ’(’ formal_parameter { ’;’ formal_parameter } ’)’ ] ’:’ parameter_type ’;’ .
procedure_head = PROCEDURE procedure_id [ ’(’ [ VAR ] formal_parameter { ’;’ [ VAR ] formal_parameter } ’)’ ] ’;’ .

actual_parameter_list = ’(’ parameter { ’,’ parameter } ’)’ .
function_call = ( built_in_function | function_ref ) [ actual_parameter_list ] .
procedure_call_stmt = ( built_in_procedure | procedure_ref ) [ actual_parameter_list ] ’;’ .
```

Are they valid or not?

_Originally posted by @zakjan in https://github.com/lutaml/expressir/issues/11#issuecomment-729624529_

**Comment by ronaldtse** [comment]
Screenshot for 9.5.2.

<img width="563" alt="Screen Shot 2020-11-19 at 3 19 21 PM" src="https://user-images.githubusercontent.com/11865/99633897-9b036600-2a7a-11eb-93b8-540fe9bd24c9.png">

**Comment by ronaldtse** [comment]
Reported at: http://bz.pdes-ch.org/show_bug.cgi?id=8482

**Comment by ronaldtse** [comment]
From Robert Swindells: http://bz.pdes-ch.org/show_bug.cgi?id=8482#c1

>> ```
>> emptyProcedure;
>> ```

> The error is thrown by the call to the procedure, not the declaration of it.

**Comment by ronaldtse** [comment]
My response:

Thanks Robert! In this case, what should the correct "procedure call" format be?

According to ISO 10303-11, 13.8, the syntax given is:
```
270 procedure_call_stmt = ( built_in_procedure | procedure_ref )
    [ actual_parameter_list ] ’;’ .
167 actual_parameter_list = ’(’ parameter { ’,’ parameter } ’)’ .
264 parameter = expression .
```

"actual_parameter_list" is optional.

In 13.8 the example given is:
```
INSERT (point_list, this_point, here );
```

**Comment by zakjan** [comment]
Yes, a declaration without params is valid, a call without params throws.

**Comment by ronaldtse** [comment]
Verified fixed on current `main` with the parsanol-based parser.

All cases from this issue parse cleanly:
- Alias statements with nested qualifiers ✓
- Binary literals in constants (`%01111000...`) ✓
- Encoded string literals ("`0000078...`") ✓
- Procedures and functions with empty parameter lists ✓

The original `eengine` parser (Stefan Cyrus's CL-based implementation) is no longer used; these issues were clarification requests against that implementation. Closing as obsolete.
